core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToStackConverter.java

```diff
@@ -335,6 +344,40 @@ private void validateStackAuthentication(StackV4Request source) {
         return convertedSet;
     }
 
+    private Set<LoadBalancer> createLoadBalancers(StackV4Request source, Stack stack) {
+        Set<LoadBalancer> loadBalancers = new HashSet<>();
+        Set<TargetGroup> targetGroups = new HashSet<>();
+        // TODO expand this to data hubs
+        if (StackType.DATALAKE.equals(source.getType())) {
+            LOGGER.info("Setting up load balancers for stack {}", source.getName());
```

Comment:
**question:** The body of this `if` is adding knox TargetGroups to the targetGroups set if the stack we're working with is a Datalake?

If possible, I think it would be clearer to move the knoxGatewayGroups out of the if block and do an else:

```java
    private Set<LoadBalancer> createLoadBalancers(StackV4Request source, Stack stack) {
        Set<LoadBalancer> loadBalancers = new HashSet<>();
        Set<TargetGroup> targetGroups = new HashSet<>();
        // TODO expand this to data hubs

        Set<String> knoxGatewayGroupNames = loadBalancerConfigService.getKnoxGatewayGroups(stack);
        Set<InstanceGroup> knoxGatewayGroups = stack.getInstanceGroups().stream()
                .filter(ig -> knoxGatewayGroupNames.contains(ig.getGroupName()))
                .collect(Collectors.toSet());

        if (StackType.DATALAKE.equals(source.getType()) && !knoxGatewayGroups.isEmpty()) {
            LOGGER.info("Setting up load balancers for stack {}", source.getName());
            LOGGER.info("Knox gateway instance found; enabling Knox load balancer configuration.");
            TargetGroup targetGroup = new TargetGroup();
            targetGroup.setType(TargetGroupType.KNOX.name());
            targetGroup.setInstanceGroups(knoxGatewayGroups);
            targetGroups.add(targetGroup);
            // TODO create target group for CM instances
        } else {
            LoadBalancer loadBalancer = new LoadBalancer();
            loadBalancer.setStack(stack);
            // TODO actually figure out the type we need here
            loadBalancer.setType(LoadBalancerType.PRIVATE.name());
            loadBalancer.setTargetGroups(targetGroups);
            targetGroups.forEach(tg -> tg.setLoadBalancer(loadBalancer));
            loadBalancers.add(loadBalancer);
        }

        return loadBalancers;
    }
```

Reply:
My intention was to limit the stack types that could have load balancers at all to only data lakes. If it's any other kind of stack we shouldn't add anything to the target groups at all. So having an `else` at all is anathema to that. It's a gatekeeping `if`.

The way target groups work is they're just a type of traffic and the instance groups that traffic is forwarded to. In the `else` you're creating target groups, but not setting either of those properties. Which is basically saying to the load balancer, "I want you to forward traffic, but I'm not going to tell you what traffic to forward or where it's going." Which is making me realize I should really mark those fields as non-null so we know they need to be set.

With that in mind, do you have any suggestions on how to clarify it?